web/src/beta/features/Visualizer/Crust/Widgets/Widget/builtin/GoogleMapSearch/index.tsx

```diff
@@ -1,34 +1,394 @@
+            }) => Promise<{
+              places: Places[];
+            }>;
+          };
+        };
+      };
+    };
+  }
+}
+
 type Property = CommonBuiltInWidgetProperty & {
   default?: {
     apiToken?: string;
   };
+  languageSetting?: {
+    language?: string;
+  };
 };
 type GoogleMapSearchProps = WidgetProps<Property>;
 
-const GoogleMapSearch: FC<GoogleMapSearchProps> = ({ widget }) => {
+type GooglePlace = {
+  place_id: string;
+  formatted_address: string;
+  name: string;
+  geometry: {
+    location: {
+      lat: number;
+      lng: number;
+    };
+  };
+  layerId?: string;
+};
+
+let loaderInstance: Loader | null = null;
+
+const removeGoogleMapsScript = () => {
+  const scripts = document.querySelectorAll<HTMLScriptElement>(
+    'script[src*="maps.googleapis.com/maps/api/js"]'
+  );
+  scripts.forEach((script) => {
+    script.remove();
+  });
+  if (window.google) {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-expect-error
+    delete window.google;
+  }
+  (Loader as any).instance = null;
+  (Loader as any).DONE = false;
+  (Loader as any).FAILED = false;
+};
+
+function getLoaderInstance(apiKey: string, language: string): Loader {
+  removeGoogleMapsScript();
+  loaderInstance = new Loader({
+    apiKey,
+    libraries: ["places"],
+    language
+  });
+
+  return loaderInstance;
+}
+
+const GoogleMapSearch: FC<GoogleMapSearchProps> = ({
+  widget,
+  context: { onFlyTo } = {}
+}) => {
   const theme = useMemo(
     () => widget.property?.appearance?.theme ?? "light",
     [widget.property?.appearance?.theme]
   );
+  const apiToken = useMemo(
+    () => widget?.property?.default?.apiToken,
+    [widget?.property?.default?.apiToken]
+  );
+  const language = useMemo(() => {
+    const userLang = widget?.property?.languageSetting?.language;
+    return userLang && userLang !== "auto" ? userLang : navigator.language;
+  }, [widget?.property?.languageSetting?.language]);
+
+  const [query, setQuery] = useState("");
+  const [debouncedQuery, setDebouncedQuery] = useState(query);
+  const [filteredSuggestions, setFilteredSuggestions] = useState<GooglePlace[]>(
+    []
+  );
+  const [selectedItems, setSelectedItems] = useState<GooglePlace[]>([]);
+  const [selectedItemIndex, setSelectedItemIndex] = useState<number | null>(
+    null
+  );
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState<string | null>(null);
+  const [googleLoaded, setGoogleLoaded] = useState(false);
+
+  const VisualizerRef = useVisualizer();
+
+  useEffect(() => {
+    if (!apiToken || !language) {
+      return;
+    }
+    setError(null);
+    const loader = getLoaderInstance(apiToken, language);
+    loader
+      .load()
+      .then(() => {
+        setGoogleLoaded(true);
+      })
+      .catch((err) => {
+        setError("Failed to load Google Maps JS API: " + err);
+      });
+  }, [apiToken, language]);
+
+  useEffect(() => {
+    if (filteredSuggestions) {
```

Comment:
```suggestion
    if (filteredSuggestionsRef.current) {
```

const filteredSuggestionsRef = useRef(filteredSuggestions);
filteredSuggestionsRef.current = filteredSuggestions;